### DISCUSSION TITLE
Query Regarding Packet Schedulers in OMNeT++

### DISCUSSION THREAD

**Discussion by ahmed3009** [omnetpp/omnetpp#1016] [Q&A] [answered]
HI all,
I had a question regarding the implementation of Packet schedulers in INET/OMNeT++. It seems the current implementation always tries to schedule a packet. e.g. in the case of WRR scheduler, even if there are insufficient tokens, the "firstWeighted" parameter makes sure that at least one of the packet providers is selected each time. But in reality, if none of the providers have sufficient tokens, and in a scheme such as DRR* (Deficit Round Robin) which relies on increasing tokens each round, there is a case where none of the providers may posses sufficient tokens to schedule a packet. In this case, such packet(s), should for example stay queued until the next round or until another packet arrives in another queue/provider with sufficient tokens. This scenario seems to be missing and I am unable to not schedule any packet in a particular round. Is there a way to cause the packet scheduler to not schedule a packet? Because by returning a gate index = -1 or a gate index which corresponds to a provider that does not have a packet results in an error.
Any thoughts or suggestions on this?

*ref =  Shreedhar, M.; Varghese,G. (October 1995). "Efficient fair queueing using deficit round robin". ACM SIGCOMM Computer Communication Review. 25 (4): 231. or https://en.wikipedia.org/wiki/ Deficit_round_robin

**Comment by levy** [comment]
There's a canPullPacket method in the IPassivePacketSource interface for exactly this purpose. The scheduler should return nullptr if all of its inputs return nullptr, effectively preventing to pull a packet from it.

**Reply by ahmed3009** [reply]
I understand this case. The case you are highlighting is one where there are no packets to schedule. So, no packet can be pulled. But there can be a case where there is a packet to be pulled but e.g. the tokens associated with a particular source are insufficient at this time. The scheduler should be prevented to pull a packet in this case too. But as far as I can understand, if the schedulPacket function has been triggered and there is a packet to be scheduled, the scheduler demands a gate index regardless of whether the packet can be actually scheduled or not.

**Reply by levy** [reply, marked as answer]
Ah, I see. The WrrScheduler doesn't take the number of tokens into account in canPullPacket. The reason is that the WrrScheduler doesn't override the default behavior that is inherited from the base class. However, the scheduler could take the number of tokens into account by overriding the default behavior. Of course, in this case other methods may also have to be overridden (e.g. getNumPackets).
